src/Daum.h

```diff
@@ -0,0 +1,133 @@
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+#ifndef _GC_Daum_h
+#define _GC_Daum_h 1
+
+#include <QDebug>
+#include <QMutex>
+#include <QString>
+#include <QThread>
+#include <QTimer>
+#if defined(Q_OS_WINDOWS) || defined(Q_OS_MAC) || defined(Q_OS_LINUX)
+#include <QSerialPort>
+#endif
+
+/*
+ * This class add support for some Daum devices that are connected
+ * to the PC via serial port.
+ * This work is based on the documents found on the daum-electronic site:
+ * http://www.daum-electronic.de/de/support/supp02.html
+ * and the implementation details are inspired by the Fortius class.
+ */
+
+class Daum : public QThread {
+    Q_OBJECT
+  public:
+    // default load
+    const int kDefaultLoad = 100;
+    const int kQueryIntervalMS = 1000;
+
+    Daum(QObject *parent, QString device, QString profile);
+
+    int start();
+    int restart();
+    int pause();
+    int stop();
+    int quit();
+
+    bool discover(QString dev);
+
+    void setLoad(double load);
+    double getPower() const;
+    double getSpeed() const;
+    double getCadence() const;
+    double getHeartRate() const;
+
+  private:
+    void run();
+
+    bool openPort(QString dev);
+    bool closePort();
+    void initializeConnection();
+    bool configureForCockpitType(int cockpitType);
+    bool configureFromProfile();
+
+    bool ResetDevice();
+    bool StartProgram(unsigned int prog);
+    bool StopProgram(unsigned int prog);
+    int GetAddress();
+    int CheckCockpit();
+    int GetDeviceVersion();
+    bool SetProgram(unsigned int prog);
+    bool SetDate();
+    bool SetTime();
+    void PlaySound();
+
+    QByteArray WriteDataAndGetAnswer(QByteArray const &dat, int response_bytes);
+    char MapLoadToByte(unsigned int load) const;
+    bool isPaused() const;
+
+    // a lock for our private vars
+    mutable QMutex pvars;
+
+    QTimer *timer_;
+    QString serialDeviceName_;
+#if defined(Q_OS_WINDOWS) || defined(Q_OS_MAC) || defined(Q_OS_LINUX)
+    QSerialPort *serial_dev_;
+#else
+    QObject *serial_dev_;
+#endif
+
+    char deviceAddress_;
+    unsigned int maxDeviceLoad_;
+    unsigned int serialWriteDelay_;
+    bool playSound_;
+
+    // state
+    bool paused_;
+
+    // inbound
+    volatile int devicePower_;
+    volatile double deviceHeartRate_;
+    volatile double deviceCadence_;
+    volatile double deviceSpeed_;
+
+    // outbound
+    volatile int load_, loadToWrite_;
+    const bool forceUpdate_;
+    const QString profile_;
+
+    enum CockpitType {
```

Comment:
The cockpit types are defined in hex, while the bike returns the value in decimal. Changing to decimal values seemed to solve the communication issue between qdomyos and Daum 8008 bike

For example, according to [documentation](https://daum-electronic.de/wp-content/uploads/2022/02/interface_2003_offiz191203.pdf) a 8008 cockpit is value 40 and not 0x40 (=64). 
Values are given as following in the Daum serial documentation: 
> 10=Cardio 
20=Fitness 
30=Vita De Luxe 
40=8008 
50=8080 
60=Therapie

Reply:
Cut out from logs before changing to decimal:
```
lör jan. 28 12:20:41 2023 1674904841127 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Speed: 0"
lör jan. 28 12:20:41 2023 1674904841128 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Distance: 0"
lör jan. 28 12:20:41 2023 1674904841128 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Cadence: 0"
lör jan. 28 12:20:41 2023 1674904841129 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Watt: 0"
lör jan. 28 12:20:41 2023 1674904841129 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Heart: 0"
lör jan. 28 12:20:41 2023 1674904841176 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Speed: 0"
lör jan. 28 12:20:41 2023 1674904841176 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Distance: 0"
lör jan. 28 12:20:41 2023 1674904841177 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Cadence: 0"
lör jan. 28 12:20:41 2023 1674904841756 Debug: ../src/Daum.cpp void Daum::initializeConnection() CheckCockpit() returned  -54
lör jan. 28 12:20:41 2023 1674904841756 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Watt: 0"
lör jan. 28 12:20:41 2023 1674904841757 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Heart: 0"
lör jan. 28 12:20:41 2023 1674904841758 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Speed: 0"
lör jan. 28 12:20:41 2023 1674904841758 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Distance: 0"
lör jan. 28 12:20:41 2023 1674904841759 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Cadence: 0"
lör jan. 28 12:20:41 2023 1674904841759 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Watt: 0"
lör jan. 28 12:20:41 2023 1674904841759 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Heart: 0"
lör jan. 28 12:20:41 2023 1674904841760 Debug: ../src/virtualbike.cpp void virtualbike::bikeProvider() "bikeProvider" 0 2000 false
lör jan. 28 12:20:41 2023 1674904841761 Debug: ../src/virtualbike.cpp void virtualbike::bikeProvider() "virtual bike bluetooth not connected"
lör jan. 28 12:20:41 2023 1674904841810 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Speed: 0"
lör jan. 28 12:20:41 2023 1674904841811 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Distance: 0"
lör jan. 28 12:20:41 2023 1674904841812 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Cadence: 0"
lör jan. 28 12:20:41 2023 1674904841812 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Watt: 0"
lör jan. 28 12:20:41 2023 1674904841813 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Heart: 0"
lör jan. 28 12:20:41 2023 1674904841875 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Speed: 0"
lör jan. 28 12:20:41 2023 1674904841875 Warning: ../src/Daum.cpp void Daum::initializeConnection() unable to identify daum cockpit type 28
lör jan. 28 12:20:41 2023 1674904841876 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Distance: 0"
lör jan. 28 12:20:41 2023 1674904841885 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Cadence: 0"
lör jan. 28 12:20:41 2023 1674904841893 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Watt: 0"
lör jan. 28 12:20:41 2023 1674904841957 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Heart: 0"
lör jan. 28 12:20:41 2023 1674904841959 Debug: ../src/virtualbike.cpp void virtualbike::bikeProvider() "bikeProvider" 0 2000 false
lör jan. 28 12:20:41 2023 1674904841959 Debug: ../src/virtualbike.cpp void virtualbike::bikeProvider() "virtual bike bluetooth not connected"
lör jan. 28 12:20:41 2023 1674904841964 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Speed: 0"
lör jan. 28 12:20:41 2023 1674904841964 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Distance: 0"
lör jan. 28 12:20:41 2023 1674904841976 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Cadence: 0"
lör jan. 28 12:20:41 2023 1674904841977 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Watt: 0"
lör jan. 28 12:20:41 2023 1674904841977 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Heart: 0"
lör jan. 28 12:20:41 2023 1674904841978 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Speed: 0"
lör jan. 28 12:20:41 2023 1674904841978 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Distance: 0"
lör jan. 28 12:20:41 2023 1674904841979 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Cadence: 0"
lör jan. 28 12:20:41 2023 1674904841979 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Watt: 0"
lör jan. 28 12:20:41 2023 1674904841979 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Heart: 0"
```

Reply:
Cut out from logs after changing to decimal values:
```
lör jan. 28 13:16:59 2023 1674908219816 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Speed: 0"
lör jan. 28 13:16:59 2023 1674908219817 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Distance: 0"
lör jan. 28 13:16:59 2023 1674908219817 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Cadence: 0"
lör jan. 28 13:16:59 2023 1674908219817 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Watt: 0"
lör jan. 28 13:16:59 2023 1674908219818 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Heart: 0"
lör jan. 28 13:17:00 2023 1674908220443 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Speed: 0"
lör jan. 28 13:17:00 2023 1674908220443 Debug: ../src/Daum.cpp void Daum::initializeConnection() CheckCockpit() returned  -54
lör jan. 28 13:17:00 2023 1674908220443 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Distance: 0"
lör jan. 28 13:17:00 2023 1674908220444 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Cadence: 0"
lör jan. 28 13:17:00 2023 1674908220444 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Watt: 0"
lör jan. 28 13:17:00 2023 1674908220445 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Heart: 0"
lör jan. 28 13:17:00 2023 1674908220446 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Speed: 0"
lör jan. 28 13:17:00 2023 1674908220446 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Distance: 0"
lör jan. 28 13:17:00 2023 1674908220447 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Cadence: 0"
lör jan. 28 13:17:00 2023 1674908220447 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Watt: 0"
lör jan. 28 13:17:00 2023 1674908220447 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Heart: 0"
lör jan. 28 13:17:00 2023 1674908220448 Debug: ../src/virtualbike.cpp void virtualbike::bikeProvider() "bikeProvider" 0 2000 false
lör jan. 28 13:17:00 2023 1674908220449 Debug: ../src/virtualbike.cpp void virtualbike::bikeProvider() "virtual bike bluetooth not connected"
lör jan. 28 13:17:00 2023 1674908220494 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Speed: 0"
lör jan. 28 13:17:00 2023 1674908220495 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Distance: 0"
lör jan. 28 13:17:00 2023 1674908220495 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Cadence: 0"
lör jan. 28 13:17:00 2023 1674908220496 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Watt: 0"
lör jan. 28 13:17:00 2023 1674908220496 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Heart: 0"
lör jan. 28 13:17:00 2023 1674908220542 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Speed: 0"
lör jan. 28 13:17:00 2023 1674908220543 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Distance: 0"
lör jan. 28 13:17:00 2023 1674908220543 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Cadence: 0"
lör jan. 28 13:17:00 2023 1674908220543 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Watt: 0"
lör jan. 28 13:17:00 2023 1674908220558 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Heart: 0"
lör jan. 28 13:17:00 2023 1674908220558 Debug: ../src/Daum.cpp void Daum::initializeConnection() Daum cockpit type:  28
lör jan. 28 13:17:00 2023 1674908220558 Debug: ../src/Daum.cpp void Daum::initializeConnection()   Using communication delay: 50 msec
lör jan. 28 13:17:00 2023 1674908220559 Debug: ../src/Daum.cpp void Daum::initializeConnection()   Playing sound: true
lör jan. 28 13:17:00 2023 1674908220622 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Speed: 0"
lör jan. 28 13:17:00 2023 1674908220622 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Distance: 0"
lör jan. 28 13:17:00 2023 1674908220684 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Cadence: 0"
lör jan. 28 13:17:00 2023 1674908220802 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Watt: 25"
lör jan. 28 13:17:00 2023 1674908220864 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Heart: 0"
lör jan. 28 13:17:00 2023 1674908220925 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Speed: 0"
lör jan. 28 13:17:00 2023 1674908220926 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Distance: 0"
lör jan. 28 13:17:00 2023 1674908220984 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Cadence: 0"
lör jan. 28 13:17:00 2023 1674908220985 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Watt: 25"
lör jan. 28 13:17:00 2023 1674908220985 Debug: ../src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Heart: 0"
```

Also discovered by Zwift over Direct Connect, and power + cadence data was visible. Unfortunately, I couldn't test more since I'm not having a Zwift subscription atm.

Reply:
for zwift you have 30km for unsubscribed user if they didn't change anything

Reply:
> The cockpit types are defined in hex, while the bike returns the value in decimal. Changing to decimal values seemed to solve the communication issue between qdomyos and Daum 8008 bike
> 
> 
> 
> For example, according to [documentation](https://daum-electronic.de/wp-content/uploads/2022/02/interface_2003_offiz191203.pdf) a 8008 cockpit is value 40 and not 0x40 (=64). 
> 
> Values are given as following in the Daum serial documentation: 
> 
> > 10=Cardio 
> 
> 20=Fitness 
> 
> 30=Vita De Luxe 
> 
> 40=8008 
> 
> 50=8080 
> 
> 60=Therapie

strange this module was from golden cheatah...anyway good that it's working in this way, i will change the code reflecting your changes

Reply:
Indeed, but I've already used the free km this month. Will get 30km more at 1 Feb :)

Reply:
hahaha ok :)

Reply:
@johanlovgren I did it. Could you pull and test it again? 
Also what about the 

`COCKPIT_8008_TRS = 0x2a,` ?

How can convert this?

Reply:
yes i leaved that way for now. Let me know